### ISSUE TITLE
A parameter is not allowed in this location. Ensure that the '@' sign is in a valid location or that parameters are valid at all in this SQL statement

### ISSUE THREAD

**Issue by malikhassan85** [ErikEJ/EntityFramework.SqlServerCompact#545]
When you define a variable from certain primitive types (double, decimal for example), and then try to assign this variable to a property in your select part from the query, you get this exception:

```
Exception:

A parameter is not allowed in this location. Ensure that the '@' sign is in a valid location or that parameters are valid at all in this SQL statement

Stack trace:

  at System.Data.SqlServerCe.SqlCeCommand.CompileQueryPlan()
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteCommand(CommandBehavior behavior, String method, ResultSetOptions options)
   at System.Data.SqlServerCe.SqlCeCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.ExecuteReader()
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Microsoft.EntityFrameworkCore.Storage.Internal.NoopExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)

```

After some investigation, I found that this issue is due to the fact that the `DBType` isn't defined for the `RelationalTypeMappingParameters` that will be initiated for these primitive types. To confirm that, in the `SqlCeTypeMappingSource` class in the source code, I specified the `DBType` in the initialisation for `_decimal` private member, and the problem was solved:

```
 private readonly DecimalTypeMapping _decimal
            = new SqlCeDecimalTypeMapping("decimal(18, 2)", DbType.Decimal, 18, 2);
``` 
**Note**: You will get the same exception if you tried to run a query using `SqlCommand` with a parameter in the columns list that hasn't  a specified DbType, this example will throw the exception:

```

var ceConnectionString = ConfigurationManager.AppSettings["SqlCeConnection"];
          

            string query = "SELECT RecordId, @summary FROM Employees;";

            using (var ceConnection = new SqlCeConnection(ceConnectionString))
            {
                var command = new SqlCeCommand(
                    query, ceConnection);
                var commandParam = new SqlCeParameter("summary", 10);
                

                command.Parameters.Add(commandParam);

                ceConnection.Open();
                SqlCeDataReader reader = command.ExecuteReader();
                try
                {
                    while (reader.Read())
                    {
                        Console.WriteLine(String.Format("{0}, {1}",
                            reader[0], reader[1]));
                    }
                }
                finally
                {
                    // Always call Close when done reading.
                    reader.Close();
                }
            }

```

However, this one won't throw it:

```
 var ceConnectionString = ConfigurationManager.AppSettings["SqlCeConnection"];
          

            string query = "SELECT RecordId, @summary FROM Employees;";

            using (var ceConnection = new SqlCeConnection(ceConnectionString))
            {
                var command = new SqlCeCommand(
                    query, ceConnection);
                var commandParam = new SqlCeParameter("summary", 10);
                commandParam.DbType = DbType.Decimal;

                command.Parameters.Add(commandParam);

                ceConnection.Open();
                SqlCeDataReader reader = command.ExecuteReader();
                try
                {
                    while (reader.Read())
                    {
                        Console.WriteLine(String.Format("{0}, {1}",
                            reader[0], reader[1]));
                    }
                }
                finally
                {
                    // Always call Close when done reading.
                    reader.Close();
                }
            }

```
If you believe that my conclusion for the actual reason for this issue is correct, I will be happy to create a PR for the fix.

## Steps to produce

https://github.com/avinash-phaniraj-readify/EFCoreTestBed.git

Please check out branch: ParameterIsNotAllowed

```

            using (var dc = new SampleDataContext(options.Options))
            {
                decimal amount = (from employees in dc.Employees select employees).Sum(p => p.Salary);

                var payrollView = dc.Payrolls
                    .Select(p => new
                    {
                        PayRollId = p.RecordId,
                        PayrollDate = p.Date,
                        PayrollTaxRate = p.TaxRate,
                        TotalWages = amount
                    });

                var itemsList = payrollView.ToList();

                Console.ReadLine();
            
            }
```

**Comment by ErikEJ** [comment]
Thanks or reporting this, are you interested in doing a PR?

>>  certain primitive types (double, decimal for example)

Do you have a list of the affected types??

**Comment by malikhassan85** [comment]
Happy to work on the PR.

I don't have the full list now, these are just the types that I've tried, but when I work on the PR, I will check all the other types to make the fix comprehensive.

**Comment by ErikEJ** [comment]
Fantastic! Let me know if you stumble on any blockers...

**Comment by ErikEJ** [comment]
Fixed in release 2.1.0.4

**Comment by malikhassan85** [comment]
Thanks for fixing this issue. I was working on it as well, and I want to drop your attention that we have the same issue for 
```
   private readonly FloatTypeMapping _real
            = new SqlCeFloatTypeMapping("real");
```
The fix for that should be:

```
   private readonly FloatTypeMapping _real
            = new SqlCeFloatTypeMapping("real", DbType.Single);
```

However, there is a bug in the current version of EFCore (2.1), which even with this change, the problem won't get solved till the next release which will be (2.2). The bug is in "FloatTypeMapping" class, which is the parent class of "SqlCeFloatTypeMapping" class. The constructor in the "FloatTypeMapping" isn't propagating the DbType to it's parent constructor, and that's why it will be null all the time. 

You can verify that by checking the history of "FloatTypeMapping" in EFCore project, you will find that it's fixed in (release 2.2) branch, but it wasn't in branch (release 2.1)

**Comment by ErikEJ** [comment]
Thanks, I have logged: https://github.com/ErikEJ/EntityFramework.SqlServerCompact/issues/548

**Comment by ErikEJ** [comment]
Fixed, will be in next 2.1 release
